### ISSUE TITLE
Kafka Lookups prevent peon processes from terminating on /stop from supervisor

### ISSUE THREAD

**Issue by gvsmirnov** [druid-io/druid#3346]
**Steps to reproduce:**

1. Create a [kafka indexer](http://druid.io/docs/0.9.1.1/development/extensions-core/kafka-ingestion.html)
2. Create a [kafka lookup](http://druid.io/docs/0.9.1.1/development/extensions-core/kafka-extraction-namespace.html)
3. Wait for a kafka index task to start
4. Emulate a `/stop` request sent from supervisor to worker, e.g. via
```
http post localhost:8100/druid/worker/v1/chat/index_kafka_Example_c2f17f3de1b02c1_fmcaakbh/stop
```

**Expected behavior:**
Worker successfully stops

**Actual behavior:**
Worker process remains lingering until middle manager forcefully kills the process after timeout

---

The reason for this is that when `KafkaLookupExtractorFactory` invokes `consumerConnector.shutdown()` upon receiving the stop command, it interrupts the worker thread, and thus the shutdown does not complete properly:

```
java.lang.InterruptedException
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2072) ~[?:1.7.0_101]
	at java.util.concurrent.ThreadPoolExecutor.awaitTermination(ThreadPoolExecutor.java:1468) ~[?:1.7.0_101]
	at kafka.utils.KafkaScheduler.shutdown(KafkaScheduler.scala:88) ~[kafka_2.10-0.8.2.1.jar:?]
	at kafka.consumer.ZookeeperConsumerConnector.liftedTree1$1(ZookeeperConsumerConnector.scala:195) [kafka_2.10-0.8.2.1.jar:?]
	at kafka.consumer.ZookeeperConsumerConnector.shutdown(ZookeeperConsumerConnector.scala:193) [kafka_2.10-0.8.2.1.jar:?]
	at kafka.javaapi.consumer.ZookeeperConsumerConnector.shutdown(ZookeeperConsumerConnector.scala:119) [kafka_2.10-0.8.2.1.jar:?]
	at io.druid.query.lookup.KafkaLookupExtractorFactory$2.run(KafkaLookupExtractorFactory.java:229) [druid-kafka-extraction-namespace-0.9.2-SNAPSHOT.jar:0.9.1.1]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471) [?:1.7.0_101]
	at java.util.concurrent.FutureTask.run(FutureTask.java:262) [?:1.7.0_101]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [?:1.7.0_101]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [?:1.7.0_101]
	at java.lang.Thread.run(Thread.java:745) [?:1.7.0_101]
```

This leaves a lingering `ConsumerFetcherThread` behind, and so the process does not terminate.

**Comment by gvsmirnov** [comment]
A hacky workaround is to use `shutdown()` instead of [`shutdownNow()`](https://github.com/druid-io/druid/blob/master/extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java#L301) on the executor, but that does not account for all the cases. Simply clearing the interrupted status does not work either.

I am currently working on a pull request addressing this issue, but thought to register an issue so that if anyone else faces this problem, they can look up the stack trace.

**Comment by drcrallen** [comment]
Thanks for the report @gvsmirnov. Just FYI https://github.com/druid-io/druid/blob/druid-0.9.1.1/extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java#L231 is supposed to handle this. I'm curious if something is swallowing the InterruptedException without re-setting the thread interrupt status.

An overall better solution would probably be to have the loop exit condition be on a more strict factory lifecycle state.

**Comment by drcrallen** [comment]
Do you see the log `Error cancelling future for topic [...]` in the logs?

**Comment by drcrallen** [comment]
Also, what point did you try clearing the interrupt status?

**Comment by drcrallen** [comment]
Another possible problem is that the stop has two interrupts in it. One at shutdownNow and one at https://github.com/druid-io/druid/blob/druid-0.9.1.1/extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java#L306

**Comment by gvsmirnov** [comment]
Part of the problem is that `shutdownNow()` interrupts the worker thread, but there is nobody there to consume the interrupted status, so the when `finally` block is entered, the thread interrupted flag is up. The consumer connector's `shutdown()` does not expect this, and, when attempting to `await()` on something, it throws the `InterruptedException` before managing to shut down properly.

In my case, I did not end up with the `Error cancelling future for topic [...]`, but I do see that some other cases may trigger this, including the one you just linked.

My idea is also to change the loop condition, but with the additional interrupts this calls for a bit of refactoring. I'm doing this now. I hope I don't get too deep into it :D

**Comment by drcrallen** [comment]
@gvsmirnov If you think you can get a reasonable handling of interrupt status (with tests would kick a**) then it would be worth changing shutdown to only do the interrupt once

**Comment by gvsmirnov** [comment]
@drcrallen Yep, I think I can do that. However, I think I accidentally got carried away. I started just extracted one method, and the next thing I know is this: https://github.com/druid-io/druid/compare/master...Plumbr:kafka-lookups?expand=1

The functionality was not changed, this was only meant to increase maintainability. The main pain points were:

* Splitting up the `start()` method into smaller methods
* Removing code duplication
* Incapsulating internal state better, and avoiding directly depending on it in unit tests

But now I realized that this is quite a large overhaul to accept from an unknown person. Should I continue refactoring the kafka lookups in this spirit? Or do you think there's little chance of getting something like that merged, and it would be better to scratch it and simply fix the bug, without addressing any of the non-functional issues?

**Comment by drcrallen** [comment]
@gvsmirnov its a modification to only one file in an extension that is still considered experimental. It's totally clobber-able by new collaborators. I try to write as extensive unit tests as I can think of just for such things.

The criteria I'll be considering in the review is:

1. Does it fix the problem
2. Does it prove that it fixes the problem (unit tests and "real" test by reporter if possible)
3. Does it retain prior functionality
4. Is it maintainable (well formatted, correctly named variables, adheres to both explicit and implicit code standards, etc)

**Comment by drcrallen** [comment]
As a general FYI though, if you have more ideas for fixes be sure to declare them in advance, and also try to minimize individual patches. 5 PRs at 500 lines is much better than a 2500 line patch.

**Comment by drcrallen** [comment]
But sometimes splitting things out is not feasible.

**Comment by gvsmirnov** [comment]
Got that, thank you for the detailed explanation. I'll go on with these guidelines in mind, and submit a PR soon.

**Comment by gianm** [comment]
fixed by #3403, according to that issue's top comment.
